In [1]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
%matplotlib inline

## for feature slection

from sklearn.linear_model import Lasso
from sklearn.feature_selection import SelectFromModel

# to visualise al the columns in the dataframe
pd.pandas.set_option('display.max_columns', None)

In [2]:
df7=pd.read_csv('X_train.csv')

In [3]:
df7.head()

SEVERITYCODE  OBJECTID  INCKEY  COLDETKEY  SEVERITYCODE.1  PERSONCOUNT  \
0             2         1    1307       1307               2            2   
1             1         2   52200      52200               1            2   
2             1         3   26700      26700               1            4   
3             1         4    1144       1144               1            3   
4             2         5   17700      17700               2            2   

   PEDCOUNT  PEDCYLCOUNT  VEHCOUNT  SDOT_COLCODE  SEGLANEKEY  CROSSWALKKEY  \
0         0            0         2            11           0             0   
1         0            0         2            16           0             0   
2         0            0         3            14           0             0   
3         0            0         3            11           0             0   
4         0            0         2            11           0             0   

   STATUS_Matched  STATUS_Unmatched  ADDRTYPE_Block  ADDRTYPE_Intersection  \
0               1                 0               0                      1   
1               1                 0               1                      0   
2               1                 0               1                      0   
3               1                 0               1                      0   
4               1                 0               0                      1   

   ADDRTYPE_Rare_var  EXCEPTRSNCODE_   EXCEPTRSNCODE_Missing  \
0                  0                1                      0   
1                  0                0                      1   
2                  0                0                      1   
3                  0                1                      0   
4                  0                0                      1   

   EXCEPTRSNCODE_NEI  SEVERITYDESC_Injury Collision  \
0                  0                              1   
1                  0                              0   
2                  0                              0   
3                  0                              0   
4                  0                              1   

   SEVERITYDESC_Property Damage Only Collision  COLLISIONTYPE_Angles  \
0                                            0                     1   
1                                            1                     0   
2                                            1                     0   
3                                            1                     0   
4                                            0                     1   

   COLLISIONTYPE_Cycles  COLLISIONTYPE_Head On  COLLISIONTYPE_Left Turn  \
0                     0                      0                        0   
1                     0                      0                        0   
2                     0                      0                        0   
3                     0                      0                        0   
4                     0                      0                        0   

   COLLISIONTYPE_Missing  COLLISIONTYPE_Other  COLLISIONTYPE_Parked Car  \
0                      0                    0                         0   
1                      0                    0                         0   
2                      0                    0                         1   
3                      0                    1                         0   
4                      0                    0                         0   

   COLLISIONTYPE_Pedestrian  COLLISIONTYPE_Rear Ended  \
0                         0                         0   
1                         0                         0   
2                         0                         0   
3                         0                         0   
4                         0                         0   

   COLLISIONTYPE_Right Turn  COLLISIONTYPE_Sideswipe  INCDATE_Rare_var  \
0                         0                        0                 1   
1                         0                        1        

In [4]:
df7.shape

(194673, 118)

In [5]:
df8 = df7.copy()

In [6]:
df7 = df7.drop(columns = ['SEVERITYDESC_Injury Collision','SEVERITYCODE.1','SEVERITYDESC_Property Damage Only Collision'])

## Balancing the Dataset

In [7]:
from sklearn.utils import resample

In [8]:
# Seperate majority and minority classes
colData_majority = df7[df7.SEVERITYCODE==1]
colData_minority = df7[df7.SEVERITYCODE==2]

#Downsample majority class
colData_ninority_upsampled = resample(colData_minority,
                                        replace=True,
                                        n_samples=len(colData_majority),
                                        random_state=123)

# Combine minority class with downsampled majority class
colData_balanced = pd.concat([colData_ninority_upsampled, colData_majority])

# Display new class counts
colData_balanced.SEVERITYCODE.value_counts()

2    136485
1    136485
Name: SEVERITYCODE, dtype: int64

fraud_upsampled = resample(fraud,
                          replace=True, # sample with replacement
                          n_samples=len(not_fraud), # match number in majority class
                          random_state=27) # reproducible results

# combine majority and upsampled minority
upsampled = pd.concat([not_fraud, fraud_upsampled])


In [9]:
colData_balanced.head()
colData_balanced.shape

(272970, 115)

In [10]:
df9 = colData_balanced

In [11]:
X = df9.drop(columns=['SEVERITYCODE'])

In [12]:
y=df9[['SEVERITYCODE']]
y

SEVERITYCODE
177818             2
53874              2
96879              2
61004              2
183035             2
...              ...
194664             1
194665             1
194667             1
194669             1
194672             1

[272970 rows x 1 columns]

In [13]:
from sklearn.preprocessing import MinMaxScaler

In [14]:
scaler = MinMaxScaler()
scaled_data = scaler.fit_transform(X)
scaled_data = pd.DataFrame(scaled_data, columns=X.columns)
scaled_data.head()
X = scaled_data

In [15]:
X

OBJECTID    INCKEY  COLDETKEY  PERSONCOUNT  PEDCOUNT  PEDCYLCOUNT  \
0       0.909226  0.872399   0.871048     0.052632       0.0          0.0   
1       0.271897  0.218639   0.217651     0.017544       0.0          0.0   
2       0.484573  0.361522   0.359888     0.035088       0.0          0.5   
3       0.305740  0.245560   0.244450     0.017544       0.0          0.0   
4       0.937193  0.986237   0.986299     0.035088       0.0          0.5   
...          ...       ...        ...          ...       ...          ...   
272965  0.999954  0.932723   0.932364     0.140351       0.0          0.0   
272966  0.999959  0.930477   0.930129     0.035088       0.0          0.0   
272967  0.999973  0.928432   0.928092     0.017544       0.0          0.0   
272968  0.999986  0.932308   0.931951     0.035088       0.0          0.0   
272969  1.000000  0.929690   0.929345     0.035088       0.0          0.0   

        VEHCOUNT  SDOT_COLCODE  SEGLANEKEY  CROSSWALKKEY  STATUS_Matched  \
0       0.166667      0.159420    0.000000           0.0             1.0   
1       0.083333      0.405797    0.000000           0.0             1.0   
2       0.083333      0.811594    0.012015           0.0             1.0   
3       0.083333      0.420290    0.000000           0.0             1.0   
4       0.083333      0.260870    0.000000           0.0             1.0   
...          ...           ...         ...           ...             ...   
272965  0.250000      0.159420    0.000000           0.0             1.0   
272966  0.166667      0.159420    0.000000           0.0             1.0   
272967  0.083333      0.405797    0.000000           0.0             1.0   
272968  0.166667      0.202899    0.000000           0.0             1.0   
272969  0.166667      0.202899    0.000000           0.0             1.0   

        STATUS_Unmatched  ADDRTYPE_Block  ADDRTYPE_Intersection  \
0                    0.0             0.0                    1.0   
1                    0.0             1.0                    0.0   
2                    0.0             1.0                    0.0   
3                    0.0             0.0                    1.0   
4                    0.0             0.0                    1.0   
...                  ...             ...                    ...   
272965               0.0             0.0                    1.0   
272966               0.0             0.0                    1.0   
272967               0.0             1.0                    0.0   
272968               0.0             1.0                    0.0   
272969               0.0             1.0                    0.0   

        ADDRTYPE_Rare_var  EXCEPTRSNCODE_   EXCEPTRSNCODE_Missing  \
0                     0.0              1.0                    0.0   
1                     0.0              0.0                    1.0   
2                     0.0              0.0                    1.0   
3                     0.0              0.0                    1.0   
4                     0.0              1.0                    0.0   
...                   ...              ...                    ...   
272965                0.0              1.0                    0.0   
272966                0.0              1.0                    0.0   
272967                0.0              1.0                    0.0   
272968                0.0              1.0                    0.0   
272969                0.0              1.0                    0.0   

        EXCEPTRSNCODE_NEI  COLLISIONTYPE_Angles  COLLISIONTYPE_Cycles  \
0                     0.0                   0.0                   0.0   
1                     0.0                   0.0                   0.0   
2                     0.0                   0.0                   1.0   
3                     0.0                   0.0                   0.0   
4                     0.0                   0.0                   1.0   
...                   ...                   ...                   ...   
272965                0.0

In [16]:
scaler = MinMaxScaler()
scaled_data1 = scaler.fit_transform(y)
scaled_data1 = pd.DataFrame(scaled_data1, columns=y.columns)
scaled_data1.head()
y=scaled_data1 

In [17]:
y

SEVERITYCODE
0                1.0
1                1.0
2                1.0
3                1.0
4                1.0
...              ...
272965           0.0
272966           0.0
272967           0.0
272968           0.0
272969           0.0

[272970 rows x 1 columns]

In [18]:
feature_sel_model = SelectFromModel(Lasso(alpha=0.005, random_state=0)) # remember to set the seed, the random state in this function
feature_sel_model.fit(X, y)

SelectFromModel(estimator=Lasso(alpha=0.005, copy_X=True, fit_intercept=True,
                                max_iter=1000, normalize=False, positive=False,
                                precompute=False, random_state=0,
                                selection='cyclic', tol=0.0001,
                                warm_start=False),
                max_features=None, norm_order=1, prefit=False, threshold=None)

In [19]:
feature_sel_model.get_support()

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False,  True, False, False, False,  True,  True,  True,  True,
       False,  True, False, False, False,  True, False, False, False,
       False, False,  True, False, False, False, False,  True, False,
       False, False, False, False, False, False, False, False,  True,
       False, False, False, False,  True,  True, False, False,  True,
       False, False, False,  True, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False,  True, False, False, False, False, False, False,  True,
       False, False, False, False, False, False])

In [20]:
X.head(10)

OBJECTID    INCKEY  COLDETKEY  PERSONCOUNT  PEDCOUNT  PEDCYLCOUNT  \
0  0.909226  0.872399   0.871048     0.052632       0.0          0.0   
1  0.271897  0.218639   0.217651     0.017544       0.0          0.0   
2  0.484573  0.361522   0.359888     0.035088       0.0          0.5   
3  0.305740  0.245560   0.244450     0.017544       0.0          0.0   
4  0.937193  0.986237   0.986299     0.035088       0.0          0.5   
5  0.265594  0.219468   0.218477     0.035088       0.0          0.0   
6  0.138909  0.131187   0.130594     0.052632       0.0          0.0   
7  0.798903  0.679092   0.676987     0.105263       0.0          0.0   
8  0.411025  0.314390   0.312969     0.035088       0.0          0.0   
9  0.384029  0.295452   0.294117     0.070175       0.0          0.0   

   VEHCOUNT  SDOT_COLCODE  SEGLANEKEY  CROSSWALKKEY  STATUS_Matched  \
0  0.166667      0.159420    0.000000           0.0             1.0   
1  0.083333      0.405797    0.000000           0.0             1.0   
2  0.083333      0.811594    0.012015           0.0             1.0   
3  0.083333      0.420290    0.000000           0.0             1.0   
4  0.083333      0.260870    0.000000           0.0             1.0   
5  0.166667      0.202899    0.000000           0.0             1.0   
6  0.166667      0.159420    0.000000           0.0             1.0   
7  0.166667      0.159420    0.000000           0.0             1.0   
8  0.166667      0.159420    0.000000           0.0             1.0   
9  0.250000      0.202899    0.000000           0.0             1.0   

   STATUS_Unmatched  ADDRTYPE_Block  ADDRTYPE_Intersection  ADDRTYPE_Rare_var  \
0               0.0             0.0                    1.0                0.0   
1               0.0             1.0                    0.0                0.0   
2               0.0             1.0                    0.0                0.0   
3               0.0             0.0                    1.0                0.0   
4               0.0             0.0                    1.0                0.0   
5               0.0             1.0                    0.0                0.0   
6               0.0             1.0                    0.0                0.0   
7               0.0             0.0                    1.0                0.0   
8               0.0             0.0                    1.0                0.0   
9               0.0             1.0                    0.0                0.0   

   EXCEPTRSNCODE_   EXCEPTRSNCODE_Missing  EXCEPTRSNCODE_NEI  \
0              1.0                    0.0                0.0   
1              0.0                    1.0                0.0   
2              0.0                    1.0                0.0   
3              0.0                    1.0                0.0   
4              1.0                    0.0                0.0   
5              0.0                    1.0                0.0   
6              0.0                    1.0                0.0   
7              1.0                    0.0                0.0   
8              0.0                    1.0                0.0   
9              0.0                    1.0                0.0   

   COLLISIONTYPE_Angles  COLLISIONTYPE_Cycles  COLLISIONTYPE_Head On  \
0                   0.0                   0.0                    0.0   
1                   0.0                   0.0                    0.0   
2                   0.0                   1.0                    0.0   
3                   0.0                   0.0                    0.0   
4                   0.0                   1.0                    0.0   
5                   0.0                   0.0                    0.0   
6                   0.0                   0.0                    0.0   
7                   1.0                   0.0                    0.0   
8                   0.0                   0.0                    0.0   
9                   0.0                   0.0                    0.0   

   COLLISIONTYPE_Left Turn  COLLISIONT

In [21]:
# let's print the number of total and selected features

# this is how we can make a list of the selected features
selected_feat = X.columns[(feature_sel_model.get_support())]

# let's print some stats
print('total features: {}'.format((X.shape[1])))
print('selected features: {}'.format(len(selected_feat)))


total features: 114
selected features: 16


In [22]:
selected_feat

Index(['COLLISIONTYPE_Cycles', 'COLLISIONTYPE_Other',
       'COLLISIONTYPE_Parked Car', 'COLLISIONTYPE_Pedestrian',
       'COLLISIONTYPE_Rear Ended', 'COLLISIONTYPE_Sideswipe',
       'JUNCTIONTYPE_At Intersection (intersection related)',
       'SDOT_COLDESC_MOTOR VEHCILE STRUCK PEDESTRIAN',
       'SDOT_COLDESC_MOTOR VEHICLE STRUCK MOTOR VEHICLE, REAR END',
       'WEATHER_Clear', 'WEATHER_Unknown', 'ROADCOND_Dry', 'ROADCOND_Unknown',
       'LIGHTCOND_Daylight',
       'ST_COLDESC_From same direction - both going straight - one stopped - rear-end',
       'ST_COLDESC_One parked--one moving'],
      dtype='object')

In [23]:
X=X[selected_feat]

In [24]:
X.head()

COLLISIONTYPE_Cycles  COLLISIONTYPE_Other  COLLISIONTYPE_Parked Car  \
0                   0.0                  0.0                       0.0   
1                   0.0                  1.0                       0.0   
2                   1.0                  0.0                       0.0   
3                   0.0                  1.0                       0.0   
4                   1.0                  0.0                       0.0   

   COLLISIONTYPE_Pedestrian  COLLISIONTYPE_Rear Ended  \
0                       0.0                       0.0   
1                       0.0                       0.0   
2                       0.0                       0.0   
3                       0.0                       0.0   
4                       0.0                       0.0   

   COLLISIONTYPE_Sideswipe  \
0                      0.0   
1                      0.0   
2                      0.0   
3                      0.0   
4                      0.0   

   JUNCTIONTYPE_At Intersection (intersection related)  \
0                                                1.0     
1                                                0.0     
2                                                0.0     
3                                                1.0     
4                                                1.0     

   SDOT_COLDESC_MOTOR VEHCILE STRUCK PEDESTRIAN  \
0                                           0.0   
1                                           0.0   
2                                           0.0   
3                                           0.0   
4                                           0.0   

   SDOT_COLDESC_MOTOR VEHICLE STRUCK MOTOR VEHICLE, REAR END  WEATHER_Clear  \
0                                                0.0                    1.0   
1                                                0.0                    1.0   
2                                                0.0                    1.0   
3                                                0.0                    0.0   
4                                                0.0                    1.0   

   WEATHER_Unknown  ROADCOND_Dry  ROADCOND_Unknown  LIGHTCOND_Daylight  \
0              0.0           1.0               0.0                 0.0   
1              0.0           1.0               0.0                 0.0   
2              0.0           1.0               0.0                 1.0   
3              0.0           0.0               0.0                 1.0   
4              0.0           0.0               0.0                 0.0   

   ST_COLDESC_From same direction - both going straight - one stopped - rear-end  \
0                                                0.0                               
1                                                0.0                               
2                                                0.0                               
3                                                0.0                               
4                                                0.0                               

   ST_COLDESC_One parked--one moving  
0                                0.0  
1                                0.0  
2                                0.0  
3                                0.0  
4                                0.0

In [25]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=4)
print ('Train set:', X_train.shape,  y_train.shape)
print ('Test set:', X_test.shape,  y_test.shape)

Train set: (191079, 16) (191079, 1)
Test set: (81891, 16) (81891, 1)


In [26]:
from sklearn.metrics import jaccard_similarity_score
from sklearn.metrics import f1_score
from sklearn.metrics import log_loss

### Random Forest

In [27]:
from sklearn.ensemble import RandomForestClassifier
#Evaluation Metrics
import time
from sklearn.metrics import accuracy_score, log_loss, jaccard_similarity_score, classification_report
from sklearn.metrics import precision_score, recall_score, roc_curve

In [28]:
#Evaluation

model_rf = RandomForestClassifier(n_estimators=10, max_features=3, max_depth =12,criterion='entropy',random_state=0, n_jobs=-1)
model_rf.fit(X_train,y_train)
yhat_rf = model_rf.predict(X_test)
jaccard_rf = jaccard_similarity_score(y_test,yhat_rf)
c_rf = classification_report(y_test,yhat_rf)
prec_rf = precision_score(y_test, yhat_rf)
rec_rf = recall_score(y_test, yhat_rf)
print('Jaccard :',jaccard_rf,'\n',
     c_rf)

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  after removing the cwd from sys.path.
C:\Users\LENOVO\AppData\Roaming\Python\Python37\site-packages\sklearn\metrics\classification.py:635: DeprecationWarning: jaccard_similarity_score has been deprecated and replaced with jaccard_score. It will be removed in version 0.23. This implementation has surprising behavior for binary and multiclass classification tasks.
  'and multiclass classification tasks.', DeprecationWarning)


Jaccard : 0.702299397980242 
               precision    recall  f1-score   support

         0.0       0.74      0.62      0.67     40789
         1.0       0.68      0.78      0.73     41102

    accuracy                           0.70     81891
   macro avg       0.71      0.70      0.70     81891
weighted avg       0.71      0.70      0.70     81891



In [29]:
from sklearn.ensemble import RandomForestClassifier
#Evaluation Metrics
rf = RandomForestClassifier(n_estimators=400, max_depth =6,criterion='entropy',min_samples_leaf=100, n_jobs=4)
rf.fit(X_train,y_train)
rf_pred = rf.predict(X_test)
rf_accuracy = (rf.score(X_train,y_train)*100)
rf_accuracy

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  after removing the cwd from sys.path.


69.27867531230538

### decision Tree

In [30]:
from sklearn.tree import DecisionTreeClassifier
colDataTree = DecisionTreeClassifier(criterion="entropy", max_depth = 7)
colDataTree
colDataTree.fit(X_train,y_train)

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=7,
                       max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort=False,
                       random_state=None, splitter='best')

In [31]:
DTyhat = colDataTree.predict(X_test)

In [32]:
jaccard_similarity_score(y_test, DTyhat)

C:\Users\LENOVO\AppData\Roaming\Python\Python37\site-packages\sklearn\metrics\classification.py:635: DeprecationWarning: jaccard_similarity_score has been deprecated and replaced with jaccard_score. It will be removed in version 0.23. This implementation has surprising behavior for binary and multiclass classification tasks.
  'and multiclass classification tasks.', DeprecationWarning)


0.6950946990511778

In [33]:
f1_score(y_test, DTyhat, average='macro')

0.6911109120098262

### Logistic Regression

In [34]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import confusion_matrix
LR = LogisticRegression(C=6, solver='liblinear').fit(X_train,y_train)
LR

C:\Users\LENOVO\AppData\Roaming\Python\Python37\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


LogisticRegression(C=6, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='warn', n_jobs=None, penalty='l2',
                   random_state=None, solver='liblinear', tol=0.0001, verbose=0,
                   warm_start=False)

In [35]:
LRyhat = LR.predict(X_test)
yhat_prob = LR.predict_proba(X_test)

In [36]:
jaccard_similarity_score(y_test, LRyhat)

C:\Users\LENOVO\AppData\Roaming\Python\Python37\site-packages\sklearn\metrics\classification.py:635: DeprecationWarning: jaccard_similarity_score has been deprecated and replaced with jaccard_score. It will be removed in version 0.23. This implementation has surprising behavior for binary and multiclass classification tasks.
  'and multiclass classification tasks.', DeprecationWarning)


0.6956930554029136

In [37]:
f1_score(y_test, LRyhat, average='macro')

0.6908985492754757

In [38]:
from sklearn.datasets import make_classification
from sklearn.ensemble import GradientBoostingClassifier
clf = GradientBoostingClassifier(random_state=0)
clf.fit(X_train, y_train)
GradientBoostingClassifier(random_state=0)
clf.predict(X_test)
clf.score(X_test, y_test)

C:\Users\LENOVO\AppData\Roaming\Python\Python37\site-packages\sklearn\ensemble\gradient_boosting.py:1450: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.7013102782967603

In [39]:
import xgboost as xgb
model=xgb.XGBClassifier(random_state=1,learning_rate=0.01)
model.fit(X_train, y_train)
model.score(X_test,y_test)


C:\Users\LENOVO\AppData\Roaming\Python\Python37\site-packages\sklearn\preprocessing\label.py:219: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\LENOVO\AppData\Roaming\Python\Python37\site-packages\sklearn\preprocessing\label.py:252: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


0.6963891025875859

In [40]:
Xgyhat = model.predict(X_test)

In [41]:
f1_score(y_test, Xgyhat, average='macro')

0.6924687540732657

In [42]:
jaccard_similarity_score(y_test, Xgyhat)

C:\Users\LENOVO\AppData\Roaming\Python\Python37\site-packages\sklearn\metrics\classification.py:635: DeprecationWarning: jaccard_similarity_score has been deprecated and replaced with jaccard_score. It will be removed in version 0.23. This implementation has surprising behavior for binary and multiclass classification tasks.
  'and multiclass classification tasks.', DeprecationWarning)


0.6963891025875859

# 7. Results <a id='results'>
    
    
|**Algorithm**|**Jaccard**|**f1-score**|
|---------|-------|--------|
|**Decision Tree**|  0.695 | 0.691 |
|**Logistic Regression**| 0.69 | 0.69 | 
|**Random Forest**| 0.702  | 0.70 | 
|**XGboost**| 0.696  | 0.692 | 


 With no doubt the *Logistic Regression* is the best model, because it is fast, less complex and yet gives results close to ensembling models like Random forest and XGboost  

## Discussion:
 In the beginning of this notebook, we had categorical data that was of type 'object'. This is not a data type that we could have fed through an algorithm, so label encoding was used to created new classes that were of type int8; a numerical data type.
 
After solving that issue we were presented with another - imbalanced data. As mentioned earlier, class 1 was nearly three times larger than class 2. The solution to this was downsampling the majority class with sklearn's resample tool. We downsampled to match the minority class exactly with 58188 values each.
 
Once we analyzed and cleaned the data, it was then fed through three ML models; Random Forest, Decision Tree,XG Boost and Logistic Regression. Although the first two are ideal for this project, logistic regression made the most sense because of its binary nature.

Evaluation metrics used to test the accuracy of our models were jaccard index, f-1 score and logloss for logistic regression. Choosing different max depth and hyperamater C values helped to improve our accuracy to be the best possible for different ML models.


## Conclusion:
##### In this study, I analyzed the relationship between severity of an accident and some characteristics which describe the situation that involved the accident. I identified the road condition,climate,lightning among the most important features that affect to the gravity of the accident. I built and compared 4 different classification models to predict whether an accident would have a high or low severity. These models can have multiple application in real life such as, Severity of a accident can be predicted in real time by using above data when an accident is reported and from there measures can be taken quickly.
